### DISCUSSION TITLE
Better control flow analyzer for loop

### DISCUSSION THREAD

**Discussion by codehz** [edubart/nelua-lang#98] [Ideas]
The problem:

```lua
local function test(value: isize): isize
  while true do
    -- some code (no break)
    if cond(value) then
      return value
    end
    -- another code (no break)
  end
end
```

got error: `error: a return statement is missing before function end`
but as we all know that it is impossible to break the loop without quit the whole function.

**Comment by edubart** [comment]
Thanks for reporting this case, I will look into the issue soon.

**Comment by codehz** [comment]
There is a small problem about the `while true`

If the loop conditions are unknown at compile time, this analysis is impossible.
However, it is too weird to determine the result of the entire control flow analysis based on whether the loop condition can be calculated at compile time. It make things inconsistency.

A possible solution is making `while true` as a special case of an infinite loop. In other cases, even if it can be determined at compile time(for example, taken from compile options), don't treat it as an infinite loop.

**Reply by edubart** [reply]
> A possible solution is making while true as a special case of an infinite loop.

I made exactly that in commit https://github.com/edubart/nelua-lang/commit/89435a790386c6e57eea0b09224235821db75257

That is, `while true` will be treated as a special case and the function will be allowed to omit the `return` in case the while loop does not have a `break` statement.
